TeX was written on a system with line-separators
and I use it on Linux (with line-terminators).
The problem is the final newline.
I work around it by doing "wterm_cr" right before exiting.
For this, "goto final_end" is convenient because I need to add wterm_cr only once.
This patch restores the "final_end" gotos which web2w removes.
In two places where they are non-local they are not restored.
In "Buffer size exceeded!" case wterm_cr is not issued because of a bug in TeX,
which uses "write_ln" instead of "write".

@@ -1021,14 +1021,10 @@
 The first line is special also because it may be read before \TeX\ has
 input a format file. In such cases, normal error messages cannot yet
 be given. The following code uses concepts that will be explained later.
-(If the \PASCAL\ compiler does not support non-local |@!goto|\unskip, the
-@^system dependencies@>
-statement `|goto exit(0)|' should be replaced by something that
-quietly terminates the program.)
 
 @<Report overflow of the input buffer, and abort@>=
 if (format_ident==0) 
-  {@+write_ln(term_out,"Buffer size exceeded!");exit(0);
+  {@+write_ln(term_out,"Buffer size exceeded!");exit(1);
 @.Buffer size exceeded@>
   } 
 else{@+cur_input.loc_field=first;cur_input.limit_field=last-1;
@@ -1885,15 +1881,9 @@
 goes to |end_of_TEX|. This is the only nontrivial |@!goto| statement in the
 whole program. It is used when there is no recovery from a particular error.
 
-Some \PASCAL\ compilers do not implement non-local |goto| statements.
-@^system dependencies@>
-In such cases the body of |jump_out| should simply be
-`|close_files_and_terminate|;\thinspace' followed by a call on some system
-procedure that quietly terminates the program.
-
 @<Error hand...@>=
 void jump_out(void)
-{@+ close_files_and_terminate(); exit(0);
+{@+ close_files_and_terminate(); wterm_cr; exit(1);
 } 
 
 @ Here now is the general |error| routine.
@@ -7165,7 +7155,7 @@
 state=new_line;start=1;index=0;line=0;name=0;
 force_eof=false;
 align_state=1000000;@/
-if (!init_terminal()) exit(0);
+if (!init_terminal()) goto final_end;
 limit=last;first=last+1; /*|init_terminal| has set |loc| and |last|*/ 
 } 
 
@@ -24317,7 +24307,7 @@
   {@+wterm_ln("Ouch---my internal constants have been clobbered!"
     "---case %d", bad);
 @.Ouch...clobbered@>
-  exit(0);
+  goto final_end;
   } 
 initialize(); /*set global variables to their starting values*/ 
 #ifdef @!INIT
@@ -24331,8 +24321,10 @@
 main_control(); /*come to life*/ 
 final_cleanup(); /*prepare for death*/ 
 close_files_and_terminate();
-ready_already=0;
-return 0; }
+final_end:
+wterm_cr;
+if (history <= warning_issued) return 0;
+else return 1; }
 
 @ Here we do whatever is needed to complete \TeX's job gracefully on the
 local operating system. The code here might come into play after a fatal
@@ -24473,9 +24465,9 @@
 {@+@<Initialize the input routines@>;
 if ((format_ident==0)||(buffer[loc]=='&')) 
   {@+if (format_ident!=0) initialize(); /*erase preloaded format*/ 
-  if (!open_fmt_file()) exit(0);
+  if (!open_fmt_file()) goto final_end;
   if (!load_fmt_file()) 
-    {@+w_close(&fmt_file);exit(0);
+    {@+w_close(&fmt_file);goto final_end;
     } 
   w_close(&fmt_file);
   while ((loc < limit)&&(buffer[loc]==' ')) incr(loc);
